ostia-operator/pkg/apicast/apicast.go

```diff
@@ -238,39 +226,145 @@ func newProbe(path string, port int32, initDelay int32, timeout int32, period in
 	}
 }
 
-//TODO - This should returna slice of generic "RateLimit" objects of multiple, mixed types
-func buildRateLimitPolicies(validSvcs []string, limits []v1alpha1.RateLimit) []FixedWindowRateLimiter {
-	var fixedLimitRules []FixedWindowRateLimiter
-	for _, v := range limits {
-		if contains(validSvcs, v.Service) {
-			switch v.Type {
-			case "fixed":
-				if v.TimeUnit == "minutes" {
-					v.TimeInterval = v.TimeInterval * 60
-				}
-				rule := FixedWindowRateLimiter{
-					Count:  v.MaxRequest,
-					Window: v.TimeInterval,
-					Key: LimiterKey{
-						Name:  v.Service,
-						Scope: v.Service,
-					},
-				}
-				fixedLimitRules = append(fixedLimitRules, rule)
-			default:
-				log.Errorf("unknown rate limit type %s for service. ignoring", v.Type, v.Service)
-			}
+func processRateLimitPolicies(limits []v1alpha1.RateLimit) PolicyChain {
+	var fixedLimiters []FixedWindowRateLimiter
+	var leakyLimiters []LeakyBucketRateLimiter
+	var connLimiters []ConnectionRateLimiter
+
+	for _, limit := range limits {
+		rl, err := parseRateLimit(limit)
+		if err != nil {
+			log.Error(fmt.Sprintf("error processing rate limit rule for %s - %s", limit.Name, err.Error()))
+			continue
 		}
 
+		switch rl.(type) {
+		case FixedWindowRateLimiter:
+			fwRl := rl.(FixedWindowRateLimiter)
+			fixedLimiters = append(fixedLimiters, fwRl)
+		case LeakyBucketRateLimiter:
+			lbRl := rl.(LeakyBucketRateLimiter)
+			leakyLimiters = append(leakyLimiters, lbRl)
+		case ConnectionRateLimiter:
+			crRl := rl.(ConnectionRateLimiter)
+			connLimiters = append(connLimiters, crRl)
+		default:
+			log.Errorf("unknown rate limit type %T. ignoring", rl)
+			continue
+		}
 	}
-	return fixedLimitRules
+
+	var config PolicyChainConfiguration
+	if len(fixedLimiters) > 0 {
+		config.FixedWindowLimiters = &fixedLimiters
+	}
+	if len(leakyLimiters) > 0 {
+		config.LeakyBucketLimiters = &leakyLimiters
+	}
+	if len(connLimiters) > 0 {
+		config.ConnectionLimiters = &connLimiters
+	}
+
+	return PolicyChain{"apicast.policy.rate_limit", config}
 }
 
-func contains(serviceNames []string, svc string) bool {
-	for _, service := range serviceNames {
-		if service == svc {
-			return true
+func parseRateLimit(rl v1alpha1.RateLimit) (interface{}, error) {
```

Comment:
I'd really try to avoid the empty interface.

Reply:
Sure. I've taken this approach based on @jmprusi comment previously around inferring the "limit type" by parsing the properties on the struct. The advantage of this approach is the it may be a simpler and nicer ux for the user. So I see two options:

1. Accept this for the benefit of the end user (if this is felt to be actually cleaner)
2. Enforce the user to provide us with the limit type explicitly in yaml and validate the properties (more work for end user but less ambiguity)

Do you see any other options or better approach to this?

Reply:
I'm not that familiar with go, but isn't this just trying to use polymorphism to parse different rate limits ?
IMO we just have parse function for every rate limit type and extract smaller functions for reuse.
Or use code generation and JSON schema to parse it and try to avoid this entirely.

Reply:
I see, so this is about the return value. Then we should agree on common interface of that return value. IMO we should be representing the rate limit internally somehow, then it somehow converts into APIcast policy configuration. We can have internal representation that covers all rate limits and knows how serialize into JSON that is different depending on its values. Thats what the k8s types and the operator framework do. They depend on "serializable" interface and thats it.

Reply:
Yeah that sounds like a good approach